diff --git a/picorv32.v b/picorv32.v
index af634b4..013f660 100644
--- a/picorv32.v
+++ b/picorv32.v
@@ -115,6 +115,8 @@ module picorv32 #(
 	input      [31:0] irq,
 	output reg [31:0] eoi,
 
+	input      [31:0] progaddr_irq,
+
 `ifdef RISCV_FORMAL
 	output reg        rvfi_valid,
 	output reg [63:0] rvfi_order,
@@ -1482,7 +1484,7 @@ module picorv32 #(
 						`debug($display("ST_RD:  %2d 0x%08x", latched_rd, latched_stalu ? alu_out_q : reg_out);)
 					end
 					ENABLE_IRQ && irq_state[0]: begin
-						current_pc = PROGADDR_IRQ;
+						current_pc = progaddr_irq;
 						irq_active <= 1;
 						mem_do_rinst <= 1;
 					end
