diff --git a/femtorv32_quark_fix.v femtorv32_quark.v
index 220b8d4..31f6c08 100644
--- a/femtorv32_quark_fix.v
+++ femtorv32_quark.v
@@ -44,8 +44,6 @@ module FemtoRV32(
    parameter RESET_ADDR       = 32'h00000000; 
    parameter ADDR_WIDTH       = 24;           
 
-   localparam ADDR_PAD = {(32-ADDR_WIDTH){1'b0}}; // 32-bits padding for addrs
-
  /***************************************************************************/
  // Instruction decoding.
  /***************************************************************************/
@@ -205,7 +203,7 @@ module FemtoRV32(
    wire [ADDR_WIDTH-1:0] loadstore_addr = rs1[ADDR_WIDTH-1:0] + 
 		   (instr[5] ? Simm[ADDR_WIDTH-1:0] : Iimm[ADDR_WIDTH-1:0]);
 
-   assign mem_addr = {ADDR_PAD, 
+   assign mem_addr = { 
 		       state[WAIT_INSTR_bit] | state[FETCH_INSTR_bit] ? 
 		       PC : loadstore_addr
 		     };
@@ -220,8 +218,8 @@ module FemtoRV32(
       /* verilator lint_on WIDTH */	       	       	       
       (isLUI               ? Uimm                 : 32'b0) |  // LUI
       (isALU               ? aluOut               : 32'b0) |  // ALUreg, ALUimm
-      (isAUIPC             ? {ADDR_PAD,PCplusImm} : 32'b0) |  // AUIPC
-      (isJALR   | isJAL    ? {ADDR_PAD,PCplus4  } : 32'b0) |  // JAL, JALR
+      (isAUIPC             ? {PCplusImm} : 32'b0) |  // AUIPC
+      (isJALR   | isJAL    ? {PCplus4  } : 32'b0) |  // JAL, JALR
       (isLoad              ? LOAD_data            : 32'b0);   // Load
 
    /***************************************************************************/
